Update EIP-2935: Move to Draft (#8166)
ethereum/EIPs: EIPS/eip-2935.md

author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (@gballet), Gajinder Singh (@g11tech), Tanishq Jasoria (@tanishqjasoria)
discussions-to: https://ethereum-magicians.org/t/eip-2935-save-historical-block-hashes-in-state/4565
status: Draft
type: Standards Track
category: Core
created: 2020-09-03
---

## Abstract

Store last 256 historical block hashes in a contract, and modify the `BLOCKHASH (0x40)` opcode to read and serve from this contract storage.

## Motivation

Currently BLOCKHASH opcode accesses history to resolve hash of the block number in EVM. However a more stateless client friendly way is to maintain and serve these hashes from state.

Although this is possible even in Merkle trie state, but Verkle trie state further allows bundling the BLOCKHASH witnesses (along with other witnesses) in an efficient manner making it worthwhile to have these in state.

A side benefit of this approach could be that it allows building/validating proofs related to last 256 ancestors directly against the current state.

## Specification

| Parameter | Value |
| - | - |
| `FORK_TIMESTAMP` | TBD |

> | `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|

Reviewer: This conflicts with the system address from EIP-4788.

| `HISTORY_SERVE_WINDOW` | `256` |

This EIP specifies for storing last `HISTORY_SERVE_WINDOW` block hashes in a ring buffer storage of `HISTORY_SERVE_WINDOW` length.

At the start of processing any block where `block.timestamp >= FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:

```python
def process_block_hash_history(block: Block, state: State):
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)

    # If this is the first fork block, add the parent's direct 255 ancestors as well
    if block.parent.timestamp < FORK_TIMESTAMP:
        ancestor = block.parent
        for i in range(HISTORY_SERVE_WINDOW - 1):

>             # stop at genesis block

Author: The wording here is confusing, I suggest the following:

```suggestion
            #  if ancestor.number == 0, this means we inserted the
            # genesis block hash in the previous iteration. Stop here.
```

            if ancestor.number == 0:

>                 break

Author: @g11tech @jochem-brouwer this seems incorrect, as if the fork happens at block 1, the genesis will not be stored, either by this method, or by the regular block operation. I think this check should be moved AFTER `state.insert_slot`

Reviewer 2: Why? If fork block is 1, then 

```
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)
```

This will be ran (so genesis hash is stored).

Reviewer 3: i think L51 makes one hop to the parent,so at number=1, L51 will hop to genesis block and save it, may be we can simplify it

            ancestor = ancestor.parent

>             state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number % HISTORY_SERVE_WINDOW, ancestor.hash)

Reviewer 3: this makes it look more analogous to L41

>             state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number % HISTORY_SERVE_WINDOW, ancestor.hash)

Author: actually, I think we should just rewrite it more clearly:

```python
for i in range(block.number-1, min(block.number-HISTORY_SERVE_WINDOW, 0)):
    state.insert_slot(HISTORY_STORAGE_ADDRESS, i % HISTORY_SERVE_WINDOW, ancestor.parent.hash)
    ancestor = ancestor.parent
```

then, the way it's implemented is left to the implementer.

>             state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number % HISTORY_SERVE_WINDOW, ancestor.hash)

Reviewer 3: i think this is correct  storing at (`i-1`) than `i`, since 'i` is ancestor

or we shouldn't use i, just `ancestor.number -1` since that relationship is more clear (we always store ancestor's parent)

```python
    state.insert_slot(HISTORY_STORAGE_ADDRESS, (ancestor.number -1) % HISTORY_SERVE_WINDOW, ancestor.parent.hash)
```

---
eip: 2935
title: Save historical block hashes in state

> author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (@gballet), Gajinder Singh (@g11tech), Tanishq Jasoria (@tanishqjasoria)
> description: store previous block hashes as storage slots of a system contract to allow for stateless execution

Author: ```suggestion
description: store previous block hashes as storage slots of a system contract to allow for stateless execution
author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (@gballet), Gajinder Singh (@g11tech), Tanishq Jasoria (@tanishqjasoria)
```

---
eip: 2935
title: Save historical block hashes in state

> author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (gballet), Gajinder Singh (@g11tech)

Author: ```suggestion
author: Vitalik Buterin (@vbuterin), Tomasz Stanczak (@tkstanczak), Guillaume Ballet (@gballet), Gajinder Singh (@g11tech)
```

| Parameter | Value |
| - | - |
| `FORK_TIMESTAMP` | TBD |
| `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
| `HISTORY_SERVE_WINDOW` | `256` |

At the start of processing any block where `block.timestamp > FORK_TIMESTAMP` (ie. before processing any transactions), 
* run `sstore(HISTORY_STORAGE_ADDRESS, block.number - 1, block.prevhash)`. 
* **additionally** if this is the first block of the fork, then store last `min(HISTORY_SERVE_WINDOW-1, block.number-1)` **blocks' prevhash** into their corresponding slots i.e persist the entire `HISTORY_SERVE_WINDOW` block hashes (up until the genesis) that can be accessed in the block.
* change the logic of the `BLOCKHASH` opcode as follows: return `sload(HISTORY_STORAGE_ADDRESS, arg)`.

> Edge cases:
> * For verkle at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
> * for verkle activated at block `1`, only genesis hash will be written at slot `0` as no additional history needs to be persisted.
> * for verkle activated at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.

Author: ```suggestion
Edge cases:

 * For verkle at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
 * for verkle activated at block `1`, only genesis hash will be written at slot `0` as no additional history needs to be persisted.
 * for verkle activated at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.
```

| Parameter | Value |
| - | - |
| `FORK_TIMESTAMP` | TBD |
| `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
| `HISTORY_SERVE_WINDOW` | `256` |

At the start of processing any block where `block.timestamp > FORK_TIMESTAMP` (ie. before processing any transactions):

>  * run `sstore(HISTORY_STORAGE_ADDRESS, block.number - 1, block.prevhash)`. 

Reviewer 2: Nit: there is currently no note on how to avoid EIP-158 state clearing. If an account has zero balance, zero nonce, no code, but storage, it is still considered dead, so any touch-action on this contract will delete the entire state. (So either pre-fund this account, or bump the nonce to one, this should be safe since no one should be able to deploy a contract here / have the keys of this contract)

>  * **additionally** if this is the first block of the fork, then store last `min(HISTORY_SERVE_WINDOW-1, block.number-1)` **blocks' prevhash** into their corresponding slots i.e persist the entire `HISTORY_SERVE_WINDOW` block hashes (up until the genesis) that can be accessed in the block.

Reviewer 2: ```suggestion
 * **additionally** if this is the first block of the fork, then store last `min(HISTORY_SERVE_WINDOW, block.number)` **blocks' prevhash** into their corresponding slots i.e persist the entire `HISTORY_SERVE_WINDOW` block hashes (up until the genesis) that can be accessed in the block.
```

I think there is an off-by-one error here. If this EIP gets activated at block 0, then at block 1 we get `min(255, 0)` (thus 0) and therefore the "last 0 **blocks' prevhash**" get written to the contract. From other changes to the EIP I get the impression that we want the genesis hash to be written as well.

In general I get the idea that the idea of the BLOCKHASH opcode does not change during the 256 block-window of the fork happening, therefore I think we have this off-by-one-error.

In case of the fork happening at block 256: we would still have access to the `BLOCKHASH` of the blocks 0-255 (so 256 blocks in total), so we need to write the latest 256 blocks (so 0-255) to the contract, not the latest 255 blocks (so 1-255) (`min(256-1, 256-1) = min(255,255) = 255)` per original).

Reviewer 3: this step is previous to additional step, so block 1 writes its prevhash i.e. 0 => genesis hash, and then no additional things get written

Reviewer 2: Yup, I missed the `additional` point

Author: it's correct, because in the general case, you have to add 255 blocks after adding the one just before (notice the **additionally** in bold).

In the case of genesis, you also get `min(255, 0) == 0` so no extra block get inserted.

I think we should switch to a pseudo-python implementation, because you're the second person to be confused by this (I was the first :D)

Author: so something like:

```python3
def process_block_hash_history(block :Block, state: State):
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, block.number-1, block.parent.hash)

    # If this is the first fork block, add the parent's direct 255 ancestors as well
    if block.parent.timestamp < FORK_TIMESTAMP:
        ancestor = block.parent
        for i in range(255):
            ancestor = ancestor.parent
            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number, ancestor.hash)
```

 * change the logic of the `BLOCKHASH` opcode as follows: return `sload(HISTORY_STORAGE_ADDRESS, arg)`.
 

Edge cases:

>  * For verkle at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.

Reviewer 2: Nit: this EIP is not specific to Verkle, so
```suggestion
 * For the fork to be activated at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
```

(also for the other examples)

>  * for verkle activated at block `1`, only genesis hash will be written at slot `0` as no additional history needs to be persisted.

Reviewer 3: ```suggestion
 * for activation at block `1`, only genesis hash will be written at slot `0` as no additional history needs to be persisted.
```

>  * for verkle activated at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.

Reviewer 3: ```suggestion
 * for activation at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.
```

| Parameter | Value |
| - | - |
| `FORK_TIMESTAMP` | TBD |
| `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
| `HISTORY_SERVE_WINDOW` | `256` |

> At the start of processing any block where `block.timestamp > FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:

Reviewer 2: ```suggestion
At the start of processing any block where `block.timestamp >= FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:
```

Current text does not match pseudocode

```python
def process_block_hash_history(block :Block, state: State):
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, block.number-1, block.parent.hash)

    # If this is the first fork block, add the parent's direct 255 ancestors as well
    if block.parent.timestamp < FORK_TIMESTAMP:
        ancestor = block.parent
        for i in range(HISTORY_SERVE_WINDOW - 1):
            # stop at genesis block
            if ancestor.number == 0:
                break

            ancestor = ancestor.parent
            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number, ancestor.hash)
```

> Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `HISTORY_SERVE_WINDOW` ancestors (up untill genesis).

Reviewer 2: ```suggestion
Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `HISTORY_SERVE_WINDOW` ancestors (up until genesis).
```

> For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp > FORK_TIMESTAMP`), switch the logic to:

Reviewer 2: ```suggestion
For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp >= FORK_TIMESTAMP`), switch the logic to:
```

```python
def resolve_blockhash(block: Block, state: State, arg: uint64):

>   assert(arg < block.number)

Reviewer 2: This does not match current `BLOCKHASH`. If `arg >= block.number` then return 0 (push 0 to stack)

Our code:

```typescript
      const number = runState.stack.pop()

      const diff = runState.interpreter.getBlockNumber() - number
      // block lookups must be within the past 256 blocks
      if (diff > BIGINT_256 || diff <= BIGINT_0) {
        runState.stack.push(BIGINT_0)
        return
      }
```

Author: yeah but that's precisely why we want to use the contract: be able to get further history than 256 blocks.

Reviewer 2: Yes, the point is about when you request the current block number hash, or a block number higher, I would assume that `assert` throws? (And thus push nothing to stack and exit call frame?)

Reviewer 3: applied the feedback

For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp > FORK_TIMESTAMP`), switch the logic to:

```python
def resolve_blockhash(block: Block, state: State, arg: uint64):
  assert(arg < block.number)
  return state.load_slot(HISTORY_STORAGE_ADDRESS, arg)
```
 
Edge cases:

 * For the fork to be activated at genesis, no history is written to the genesis state, and at the start of block `1`, genesis hash will be written as a normal operation to slot `0`.
 * for activation at block `1`, only genesis hash will be written at slot `0` as there is no additional history that needs to be persisted.
 * for activation at block `32`, block `31`'s hash will be written to slot `31` and additonal history for `0..30`'s hashes will be persisted, so all in all `0..31`'s hashes.

### EIP-158 exception

This address is currently exempt from [EIP-158](./eip-158.md) cleanup in Kaustinen Verkle Testnet but there are two ways this could be addressed before this EIP is adopted by ACD:

* Update the nonce to 1 in the fork block, or
* Deploy a contract à la [EIP-4788](./eip-4788.md) with `BLOCKHASH` opcode delegating call to this contract with appropriate args.

> While the second option looks more elegant, but it has higher complexity as well as gas consumption considerations.

Reviewer 2: ```suggestion
While the second option looks more elegant, it has a higher complexity as well as gas consumption considerations.
```

```python
def process_block_hash_history(block :Block, state: State):
    if block.timestamp >= FORK_TIMESTAMP:
        state.insert_slot(HISTORY_STORAGE_ADDRESS, block.number-1, block.parent.hash)

    # If this is the first fork block, add the parent's direct 255 ancestors as well
    if block.parent.timestamp < FORK_TIMESTAMP:
        ancestor = block.parent
        for i in range(MIN_HISTORY_SERVE_WINDOW - 1):
            # stop at genesis block
            if ancestor.number == 0:
                break

            ancestor = ancestor.parent
            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number, ancestor.hash)
```

Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `MIN_HISTORY_SERVE_WINDOW` ancestors (up until genesis).

For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp >= FORK_TIMESTAMP`), switch the logic to:

```python
def resolve_blockhash(block: Block, state: State, arg: uint64):
  # check the wrap around range

>   if arg >= block.number or arg < max(block.number - MAX_HISTORY_SERVE_WINDOW, 0)

Reviewer 2: In EVM, the requested `block.number` cannot be negative or be higher than 2**256, so just the check `arg >= block.number` (assuming that `block` is the current block we are running on) is sufficient here.

Reviewer 2: `arg` should thus also be of type `uint256`

Reviewer 3: yes we need to discuss the wrap around thing and arg sizes
